clientupdate/distsign/distsign.go

```diff
@@ -0,0 +1,228 @@
+// These keys are long-lived and prove the validity of current signing keys
+// from distsign.pub. To rotate root keys, a new client release must be
+// published, they are not rotated dynamically. There are multiple root keys in
+// different locations specifically to allow this rotation without using the
+// discarded root key for any new signatures.
+//
+// The signing public keys are fetched by the client dynamically before every
+// download and can be rotated more readily, assuming that most deployed
+// clients trust the root keys used to new fresh signing keys.
+package distsign
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+)
+
+const (
+	pemTypePrivate = "PRIVATE KEY"
+	pemTypePublic  = "PUBLIC KEY"
+)
+
+// GenerateKey generates a new key pair and encodes it as PEM.
+func GenerateKey() (priv, pub []byte, retErr error) {
+	pub, priv, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+	return pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePrivate,
+			Bytes: []byte(priv),
+		}), pem.EncodeToMemory(&pem.Block{
+			Type:  pemTypePublic,
+			Bytes: []byte(pub),
+		}), nil
+}
+
+// Signer is a single key (root or signing) that can sign arbitrary blobs.
+type Signer struct {
+	pk ed25519.PrivateKey
+}
+
+// NewSigner parses the PEM-encoded private key and creates a Signer for it.
+// The key is expected to be in the same format as returned by GenerateKey.
+func NewSigner(privKeyPath string) (*Signer, error) {
+	raw, err := os.ReadFile(privKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	k, err := parsePrivateKey(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse %q", privKeyPath)
+	}
+	return &Signer{pk: k}, nil
+}
+
+// Sign signs msg and returns the signature.
+func (s *Signer) Sign(msg []byte) []byte {
+	return ed25519.Sign(s.pk, msg)
+}
+
+// Client downloads and validates files from a distribution server.
+type Client struct {
+	roots    []ed25519.PublicKey
+	pkgsAddr string
+}
+
+// NewClient returns a new client for distribution server located at pkgsAddr,
+// and uses embedded root keys from the roots/ subdirectory of this package.
+func NewClient(pkgsAddr string) (*Client, error) {
+	return &Client{roots: roots(), pkgsAddr: pkgsAddr}, nil
+}
+
+// Download fetches a file at path srcPath from pkgsAddr passed in NewClient.
+// The file is downloaded to dstPath and its signature is validated using the
+// embedded root keys. Download returns an error if anything goes wrong with
+// the actual file download or with signature validation.
+func (c *Client) Download(srcPath, dstPath string) error {
+	// TODO(awly): don't buffer in memory
+
+	// Always fetch a fresh signing key.
+	sigPub, err := c.signingKeys()
+	if err != nil {
+		return err
+	}
+
+	srcURL := fmt.Sprintf("%s/%s", c.pkgsAddr, srcPath)
+	sigURL := srcURL + ".sig"
+
+	data, err := download(srcURL, 1<<30) // limit to 1Gb.
```

Comment:
Can we safely set an even lower limit? Maybe even as low as 128 MiB?

Reply:
I might also give that magic number a name, like `const ReasonablePackageSizeLimit = ...`.

Reply:
Set to 512Mb, to somewhat future-proof it for packages other than our current release builds.
This is intended as a fail-safe to avoid OOM from a malicious download, so I kept it generous.